File: test/sample.go
package main

import (
	"encoding/json"
	"fmt"
	"math/rand"
	"net/http"
	"sync"
	"time"
)

// Record represents a vinyl record in the collection
type Record struct {
	Title    string   `json:"title"`
	Artist   string   `json:"artist"`
	Year     int      `json:"year"`
	Genre    Genre    `json:"genre"`
	Tracks   []string `json:"tracks"`
	Duration float64  `json:"duration"`
}

// Genre is an enum-like type for music genres
type Genre string

const (
	Funk  Genre = "funk"
	Soul  Genre = "soul"
	Disco Genre = "disco"
	Jazz  Genre = "jazz"
)

// Collection manages a thread-safe vinyl collection
type Collection struct {
	mu      sync.RWMutex
	records map[string]*Record
	maxSize int
}

// NewCollection creates a new collection with a max size
func NewCollection(maxSize int) *Collection {
	return &Collection{
		records: make(map[string]*Record),
		maxSize: maxSize,
	}
}

// Add inserts a record into the collection
func (c *Collection) Add(r *Record) error {
	c.mu.Lock()
	defer c.mu.Unlock()

	if len(c.records) >= c.maxSize {
		return fmt.Errorf("collection full: max %d records", c.maxSize)
	}

	key := fmt.Sprintf("%s-%s", r.Artist, r.Title)
	c.records[key] = r
	return nil
}

// FindByGenre returns all records matching the given genre
func (c *Collection) FindByGenre(genre Genre) []*Record {
	c.mu.RLock()
	defer c.mu.RUnlock()

	var results []*Record
	for _, r := range c.records {
		if r.Genre == genre {
			results = append(results, r)
		}
	}
	return results
}

// FindByDecade returns records from a specific decade
func (c *Collection) FindByDecade(startYear int) []*Record {
	c.mu.RLock()
	defer c.mu.RUnlock()

	var results []*Record
	for _, r := range c.records {
		if r.Year >= startYear && r.Year < startYear+10 {
			results = append(results, r)
		}
	}
	return results
}

// Random picks a random record from the collection
func (c *Collection) Random() (*Record, bool) {
	c.mu.RLock()
	defer c.mu.RUnlock()

	if len(c.records) == 0 {
		return nil, false
	}

	keys := make([]string, 0, len(c.records))
	for k := range c.records {
		keys = append(keys, k)
	}

	idx := rand.Intn(len(keys))
	return c.records[keys[idx]], true
}

// Stats holds genre distribution statistics
type Stats struct {
	TotalRecords int            `json:"total_records"`
	ByGenre      map[Genre]int  `json:"by_genre"`
	AvgDuration  float64        `json:"avg_duration"`
	OldestYear   int            `json:"oldest_year"`
}

// GetStats computes collection statistics
func (c *Collection) GetStats() Stats {
	c.mu.RLock()
	defer c.mu.RUnlock()

	stats := Stats{
		TotalRecords: len(c.records),
		ByGenre:      make(map[Genre]int),
		OldestYear:   9999,
	}

	var totalDuration float64
	for _, r := range c.records {
		stats.ByGenre[r.Genre]++
		totalDuration += r.Duration
		if r.Year < stats.OldestYear {
			stats.OldestYear = r.Year
		}
	}

	if stats.TotalRecords > 0 {
		stats.AvgDuration = totalDuration / float64(stats.TotalRecords)
	}

	return stats
}

// Handler interface for pluggable request handling
type Handler interface {
	ServeCollection(w http.ResponseWriter, r *http.Request)
}

// APIServer serves the collection over HTTP
type APIServer struct {
	collection *Collection
	port       int
}

func (s *APIServer) ServeCollection(w http.ResponseWriter, r *http.Request) {
	switch r.Method {
	case http.MethodGet:
		genre := Genre(r.URL.Query().Get("genre"))
		var records []*Record
		if genre != "" {
			records = s.collection.FindByGenre(genre)
		} else {
			records = s.collection.FindByDecade(1970)
		}

		w.Header().Set("Content-Type", "application/json")
		if err := json.NewEncoder(w).Encode(records); err != nil {
			http.Error(w, err.Error(), http.StatusInternalServerError)
			return
		}

	case http.MethodPost:
		var rec Record
		if err := json.NewDecoder(r.Body).Decode(&rec); err != nil {
			http.Error(w, "invalid JSON", http.StatusBadRequest)
			return
		}
		defer r.Body.Close()

		if err := s.collection.Add(&rec); err != nil {
			http.Error(w, err.Error(), http.StatusConflict)
			return
		}

		w.WriteHeader(http.StatusCreated)
		fmt.Fprintf(w, "Added: %s by %s\n", rec.Title, rec.Artist)

	default:
		http.Error(w, "method not allowed", http.StatusMethodNotAllowed)
	}
}

func seedData(c *Collection) {
	records := []Record{
		{Title: "Innervisions", Artist: "Stevie Wonder", Year: 1973, Genre: Funk, Duration: 2640,
			Tracks: []string{"Too High", "Visions", "Living for the City"}},
		{Title: "Head Hunters", Artist: "Herbie Hancock", Year: 1973, Genre: Jazz, Duration: 2520,
			Tracks: []string{"Chameleon", "Watermelon Man"}},
		{Title: "Off the Wall", Artist: "Michael Jackson", Year: 1979, Genre: Disco, Duration: 2580,
			Tracks: []string{"Don't Stop 'Til You Get Enough", "Rock with You"}},
		{Title: "There's a Riot Goin' On", Artist: "Sly & Family Stone", Year: 1971, Genre: Soul, Duration: 2880,
			Tracks: []string{"Family Affair", "Runnin' Away"}},
	}

	for i := range records {
		if err := c.Add(&records[i]); err != nil {
			fmt.Printf("Warning: %v\n", err)
		}
	}
}

func main() {
	collection := NewCollection(1000)
	seedData(collection)

	stats := collection.GetStats()
	fmt.Printf("Collection: %d records, oldest from %d\n", stats.TotalRecords, stats.OldestYear)
	for genre, count := range stats.ByGenre {
		fmt.Printf("  %s: %d\n", genre, count)
	}

	if pick, ok := collection.Random(); ok {
		fmt.Printf("Random pick: %s by %s (%d)\n", pick.Title, pick.Artist, pick.Year)
	}

	server := &APIServer{collection: collection, port: 8080}
	http.HandleFunc("/records", server.ServeCollection)

	fmt.Printf("Listening on :%d\n", server.port)
	_ = http.ListenAndServe(fmt.Sprintf(":%d", server.port), nil)

	// Keep the linter happy
	_ = time.Now()
}

File: test/sample.rs
use std::collections::HashMap;
use std::fmt;
use std::sync::{Arc, Mutex};

/// Represents a music genre from the 70s era
#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub enum Genre {
    Funk,
    Soul,
    Disco,
    Jazz,
    Reggae,
}

impl fmt::Display for Genre {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Genre::Funk => write!(f, "Funk"),
            Genre::Soul => write!(f, "Soul"),
            Genre::Disco => write!(f, "Disco"),
            Genre::Jazz => write!(f, "Jazz"),
            Genre::Reggae => write!(f, "Reggae"),
        }
    }
}

/// A single vinyl record
#[derive(Debug, Clone)]
pub struct Record {
    pub title: String,
    pub artist: String,
    pub year: u16,
    pub genre: Genre,
    pub side_a: Vec<String>,
    pub side_b: Vec<String>,
    pub rating: Option<f32>,
}

impl Record {
    pub fn new(title: &str, artist: &str, year: u16, genre: Genre) -> Self {
        Self {
            title: title.to_string(),
            artist: artist.to_string(),
            year,
            genre,
            side_a: Vec::new(),
            side_b: Vec::new(),
            rating: None,
        }
    }

    pub fn with_tracks(mut self, side_a: Vec<&str>, side_b: Vec<&str>) -> Self {
        self.side_a = side_a.into_iter().map(String::from).collect();
        self.side_b = side_b.into_iter().map(String::from).collect();
        self
    }

    pub fn with_rating(mut self, rating: f32) -> Self {
        self.rating = Some(rating.clamp(0.0, 5.0));
        self
    }

    pub fn total_tracks(&self) -> usize {
        self.side_a.len() + self.side_b.len()
    }

    pub fn is_from_decade(&self, decade_start: u16) -> bool {
        self.year >= decade_start && self.year < decade_start + 10
    }
}

impl fmt::Display for Record {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let rating = self
            .rating
            .map(|r| format!(" [{:.1}/5]", r))
            .unwrap_or_default();
        write!(
            f,
            "{} - {} ({}, {}){}",
            self.artist, self.title, self.year, self.genre, rating
        )
    }
}

/// Errors that can occur in the crate
#[derive(Debug, thiserror::Error)]
pub enum CrateError {
    #[error("collection is full (max {max} records)")]
    CollectionFull { max: usize },

    #[error("record not found: {0}")]
    NotFound(String),

    #[error("invalid rating: {0} (must be 0.0-5.0)")]
    InvalidRating(f32),

    #[error("lock poisoned")]
    LockPoisoned,
}

/// Trait for anything that can be searched
pub trait Searchable {
    fn matches(&self, query: &str) -> bool;
}

impl Searchable for Record {
    fn matches(&self, query: &str) -> bool {
        let q = query.to_lowercase();
        self.title.to_lowercase().contains(&q)
            || self.artist.to_lowercase().contains(&q)
            || self.genre.to_string().to_lowercase().contains(&q)
    }
}

/// Thread-safe vinyl collection
pub struct Collection {
    records: Arc<Mutex<Vec<Record>>>,
    max_size: usize,
}

impl Collection {
    pub fn new(max_size: usize) -> Self {
        Self {
            records: Arc::new(Mutex::new(Vec::with_capacity(max_size))),
            max_size,
        }
    }

    pub fn add(&self, record: Record) -> Result<(), CrateError> {
        let mut records = self.records.lock().map_err(|_| CrateError::LockPoisoned)?;

        if records.len() >= self.max_size {
            return Err(CrateError::CollectionFull { max: self.max_size });
        }

        records.push(record);
        Ok(())
    }

    pub fn find<F>(&self, predicate: F) -> Result<Vec<Record>, CrateError>
    where
        F: Fn(&Record) -> bool,
    {
        let records = self.records.lock().map_err(|_| CrateError::LockPoisoned)?;
        Ok(records.iter().filter(|r| predicate(r)).cloned().collect())
    }

    pub fn search(&self, query: &str) -> Result<Vec<Record>, CrateError> {
        self.find(|r| r.matches(query))
    }

    pub fn by_genre(&self, genre: &Genre) -> Result<Vec<Record>, CrateError> {
        self.find(|r| &r.genre == genre)
    }

    pub fn by_decade(&self, start: u16) -> Result<Vec<Record>, CrateError> {
        self.find(|r| r.is_from_decade(start))
    }

    pub fn stats(&self) -> Result<CollectionStats, CrateError> {
        let records = self.records.lock().map_err(|_| CrateError::LockPoisoned)?;

        let mut by_genre: HashMap<Genre, usize> = HashMap::new();
        let mut total_rating = 0.0_f32;
        let mut rated_count = 0_usize;
        let mut oldest: Option<u16> = None;

        for record in records.iter() {
            *by_genre.entry(record.genre.clone()).or_insert(0) += 1;

            if let Some(rating) = record.rating {
                total_rating += rating;
                rated_count += 1;
            }

            oldest = Some(oldest.map_or(record.year, |o: u16| o.min(record.year)));
        }

        Ok(CollectionStats {
            total: records.len(),
            by_genre,
            avg_rating: if rated_count > 0 {
                Some(total_rating / rated_count as f32)
            } else {
                None
            },
            oldest_year: oldest,
        })
    }
}

impl Clone for Collection {
    fn clone(&self) -> Self {
        Self {
            records: Arc::clone(&self.records),
            max_size: self.max_size,
        }
    }
}

/// Statistics about a collection
#[derive(Debug)]
pub struct CollectionStats {
    pub total: usize,
    pub by_genre: HashMap<Genre, usize>,
    pub avg_rating: Option<f32>,
    pub oldest_year: Option<u16>,
}

impl fmt::Display for CollectionStats {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        writeln!(f, "Collection: {} records", self.total)?;
        for (genre, count) in &self.by_genre {
            writeln!(f, "  {}: {}", genre, count)?;
        }
        if let Some(avg) = self.avg_rating {
            writeln!(f, "  Avg rating: {:.1}/5", avg)?;
        }
        if let Some(year) = self.oldest_year {
            writeln!(f, "  Oldest: {}", year)?;
        }
        Ok(())
    }
}

/// Macro for quickly creating records
macro_rules! record {
    ($title:expr, $artist:expr, $year:expr, $genre:expr) => {
        Record::new($title, $artist, $year, $genre)
    };
    ($title:expr, $artist:expr, $year:expr, $genre:expr, rating: $r:expr) => {
        Record::new($title, $artist, $year, $genre).with_rating($r)
    };
}

fn main() {
    let collection = Collection::new(1000);

    let seed_records = vec![
        record!("Innervisions", "Stevie Wonder", 1973, Genre::Funk, rating: 4.8),
        record!("Head Hunters", "Herbie Hancock", 1973, Genre::Jazz, rating: 4.5)
            .with_tracks(vec!["Chameleon", "Watermelon Man"], vec!["Sly", "Vein Melter"]),
        record!("Off the Wall", "Michael Jackson", 1979, Genre::Disco, rating: 4.7),
        record!("There's a Riot Goin' On", "Sly & Family Stone", 1971, Genre::Soul),
        record!("Catch a Fire", "Bob Marley & The Wailers", 1973, Genre::Reggae, rating: 4.6),
        record!("Saturday Night Fever", "Bee Gees", 1977, Genre::Disco, rating: 4.2),
    ];

    for record in seed_records {
        if let Err(e) = collection.add(record) {
            eprintln!("Error adding record: {}", e);
        }
    }

    // Print stats
    match collection.stats() {
        Ok(stats) => print!("{}", stats),
        Err(e) => eprintln!("Error: {}", e),
    }

    // Search
    let query = "stevie";
    match collection.search(query) {
        Ok(results) => {
            println!("\nSearch '{}': {} results", query, results.len());
            for r in &results {
                println!("  {}", r);
            }
        }
        Err(e) => eprintln!("Search error: {}", e),
    }

    // By decade
    match collection.by_decade(1970) {
        Ok(seventies) => {
            println!("\n70s records:");
            for r in &seventies {
                println!("  {} ({} tracks)", r, r.total_tracks());
            }
        }
        Err(e) => eprintln!("Error: {}", e),
    }

    // Clone collection (shares Arc)
    let backup = collection.clone();
    let funk = backup.by_genre(&Genre::Funk).unwrap_or_default();
    println!("\nFunk records in backup: {}", funk.len());
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_record_creation() {
        let r = Record::new("Test", "Artist", 1975, Genre::Funk);
        assert_eq!(r.title, "Test");
        assert_eq!(r.year, 1975);
        assert!(r.rating.is_none());
    }

    #[test]
    fn test_rating_clamping() {
        let r = Record::new("Test", "Artist", 1975, Genre::Funk).with_rating(10.0);
        assert_eq!(r.rating, Some(5.0));
    }

    #[test]
    fn test_decade_check() {
        let r = Record::new("Test", "Artist", 1975, Genre::Funk);
        assert!(r.is_from_decade(1970));
        assert!(!r.is_from_decade(1980));
    }

    #[test]
    fn test_collection_full() {
        let col = Collection::new(1);
        col.add(Record::new("A", "B", 1970, Genre::Funk)).unwrap();
        let err = col.add(Record::new("C", "D", 1971, Genre::Soul));
        assert!(matches!(err, Err(CrateError::CollectionFull { max: 1 })));
    }

    #[test]
    fn test_search() {
        let col = Collection::new(10);
        col.add(Record::new("Innervisions", "Stevie Wonder", 1973, Genre::Funk))
            .unwrap();
        let results = col.search("stevie").unwrap();
        assert_eq!(results.len(), 1);
        assert_eq!(results[0].artist, "Stevie Wonder");
    }
}
